### PR TITLE
[VEN-2258]: add TimeManagerV5 and TimeManagerV8

### PR DESCRIPTION
**Pull request by GitGuru7**
## Description

This PR contains 2 contracts TimeMangerV5 and TimeManagerV8

- TimeManagerV5 is compatible with solidity version 0.5.16.
- TimeManagerV8 is compatible with solidity version 0.8.13.
- Tested  TimeManagerV5 functionalities.

Resolves VEN-2258

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/TimeManagerV8.sol
Comment (chechu) [inline contracts/TimeManagerV8.sol:1]: We should use this TimeManager contract as a reference: https://github.com/VenusProtocol/venus-protocol/pull/407/files#diff-db76421f4db0141fa672a2220d6d241d88748c227ab9c6a9cda10ce31a479265

File: contracts/TimeManagerV5.sol [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * @dev Function to simply retrieve block number or block timestamp
+     * @return Current block number or block timestamp
+     */
+    function getBlockNumberOrTimestamp() public view returns (uint256) {
+        return _getCurrentSlot();
+    }
+
+    /**
+     * @notice Returns the current timestamp in seconds
```
Comment (Debugger022) [inline contracts/TimeManagerV5.sol]: For internal or private functions we generally use dev.

File: contracts/TimeManagerV8.sol [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    error InvalidBlocksPerYear();
+
+    /// @notice Thrown when time based but blocks per year is provided
+    error InvalidTimeBased();
+
+    /**
+     * @param timeBased_ A boolean indicating whether the contract is based on time or block.
+     * If timeBased is true than blocksPerYear_ param is ignored as blocksOrSecondsPerYear is set to SECONDS_PER_YEAR
+     * @param blocksPerYear_ The number of blocks per year
+     * @custom:error InvalidBlocksPerYear is thrown if blocksPerYear entered is zero
```
Comment (Debugger022) [inline contracts/TimeManagerV8.sol]:
```suggestion
     * @custom:error InvalidBlocksPerYear is thrown if blocksPerYear entered is zero and timeBased is false
      * @custom:error InvalidTimeBased is thrown if blocksPerYear entered is not zero and timeBased is true
```

File: contracts/TimeManagerV8.sol [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+     * @custom:oz-upgrades-unsafe-allow constructor
+     */
+    constructor(bool timeBased_, uint256 blocksPerYear_) {
+        if (!timeBased_ && blocksPerYear_ == 0) {
+            revert InvalidBlocksPerYear();
+        }
+
+        if (timeBased_ && blocksPerYear_ != 0) {
+            revert InvalidTimeBased();
```
Comment (Debugger022) [inline contracts/TimeManagerV8.sol]: Should we rename the custom error to: `InvalidTimeBasedConfiguration`
for both versions of **TimeManager**.

File: contracts/TimeManagerV8.sol [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    bool public immutable isTimeBased;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    function() view returns (uint256) private immutable _getCurrentSlot;
+
+    /// @notice Thrown when blocks per year is invalid
+    error InvalidBlocksPerYear();
+
+    /// @notice Thrown when time based but blocks per year is provided
```
Comment (Debugger022) [inline contracts/TimeManagerV8.sol:26]:
```suggestion
    /// @notice Thrown when time based but blocks per year is zero
```

File: contracts/TimeManagerV5.sol [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+pragma solidity 0.5.16;
+
+contract TimeManagerV5 {
+    /// @dev The approximate number of seconds per year
+    uint256 public constant SECONDS_PER_YEAR = 31_536_000;
+
+    /// @notice Number of blocks per year or seconds per year
+    uint256 public blocksOrSecondsPerYear;
+
+    /// @dev Sets true when block timestamp is used
```
Comment (Debugger022) [inline contracts/TimeManagerV5.sol:11]:
```suggestion
    /// @dev Set to true when block timestamp is used
```

File: contracts/TimeManagerV8.sol (hunk 4 above)
Comment (GitGuru7) [reply contracts/TimeManagerV8.sol:26]: Error will be thrown when it is time-based and blocks per year are also provided

File: contracts/TimeManagerV5.sol (hunk 5 above)
Comment (GitGuru7) [reply contracts/TimeManagerV5.sol:11]: I think for boolean variable 'sets' will be preferred.

File: contracts/TimeManagerV8.sol (hunk 4 above)
Comment (Debugger022) [reply contracts/TimeManagerV8.sol:26]: Yeah, I mistyped.
We can say:
`/// @notice Thrown when time based but blocks per year is non zero`

**Review by Debugger022 — APPROVED** [review]

**Review by chechu — APPROVED** [review]

File: contracts/TimeManagerV5.sol [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    bool public isTimeBased;
+
```
Comment (chechu) [inline contracts/TimeManagerV5.sol:13]: Should we add a gap, given this contract will be extended?

Comment (GitGuru7) [reply contracts/TimeManagerV5.sol:13]: [e3f9df8](https://github.com/VenusProtocol/solidity-utilities/pull/8/commits/e3f9df8370b72d3f0631b2547911a96d4451385c)

**Pull request merged**